core/src/main/java/org/springframework/security/oauth2/server/authorization/authentication/OAuth2ClientCredentialsAuthenticationProvider.java

```diff
@@ -0,0 +1,66 @@
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.security.oauth2.server.authorization.authentication;
+
+import org.springframework.security.authentication.AuthenticationProvider;
+import org.springframework.security.core.Authentication;
+import org.springframework.security.core.AuthenticationException;
+import org.springframework.security.oauth2.core.OAuth2AuthenticationException;
+import org.springframework.security.oauth2.core.OAuth2Error;
+import org.springframework.security.oauth2.core.OAuth2ErrorCodes;
+import org.springframework.security.oauth2.server.authorization.client.RegisteredClient;
+import org.springframework.security.oauth2.server.authorization.client.RegisteredClientRepository;
+import org.springframework.security.oauth2.server.authorization.token.TokenIssuer;
+
+/**
+ * An {@link AuthenticationProvider} that converts give {@link OAuth2ClientCredentialsAuthenticationToken} to a {@link OAuth2AccessTokenAuthenticationToken}
+ * using the provided token generator.
+ *
+ * @author Alexey Nesterov
+ */
+public class OAuth2ClientCredentialsAuthenticationProvider implements AuthenticationProvider {
+
+	private final RegisteredClientRepository registeredClientsRepository;
+	private final TokenIssuer<RegisteredClient> tokenIssuer;
+
+	public OAuth2ClientCredentialsAuthenticationProvider(RegisteredClientRepository registeredClientsRepository, TokenIssuer<RegisteredClient> tokenIssuer) {
+		this.registeredClientsRepository = registeredClientsRepository;
+		this.tokenIssuer = tokenIssuer;
+	}
+
+	@Override
+	public Authentication authenticate(Authentication authentication) throws AuthenticationException {
+		final OAuth2ClientCredentialsAuthenticationToken token = (OAuth2ClientCredentialsAuthenticationToken) authentication;
+		final String clientId = token.getPrincipal().toString();
+		final String clientSecret = token.getCredentials().toString();
+
+		RegisteredClient registeredClient = this.registeredClientsRepository.findByClientId(clientId);
+		if (registeredClient == null) {
+			throw new OAuth2AuthenticationException(new OAuth2Error(OAuth2ErrorCodes.INVALID_CLIENT));
+		}
+
+		if (!clientSecret.equals(registeredClient.getClientSecret())) {
+			throw new OAuth2AuthenticationException(new OAuth2Error(OAuth2ErrorCodes.INVALID_CLIENT));
+		}
+
+		return new OAuth2AccessTokenAuthenticationToken(registeredClient, token, this.tokenIssuer.issue(registeredClient));
+	}
+
+	@Override
+	public boolean supports(Class<?> authentication) {
+		return authentication == OAuth2ClientCredentialsAuthenticationToken.class;
```

Comment:
Change to `return OAuth2ClientCredentialsAuthenticationToken.class.isAssignableFrom(authentication)`